### PR TITLE
8303942: FileMapInfo::write_bytes aborts on a short os::write

### PR DESCRIPTION
**Pull request by afshin-zafari**
`os::write` is called in a loop until all the requested size is written. The number of bytes parameter(`size_t nbytes`) is casted to `ssize_t size` to be able to check `> 0` condition. To enable pointer arithmetic, the `const void *` is casted to `const char *` for addition and then recasted back.

### Test
local: hotspot:tier1
mach5 tier1-5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303942](https://bugs.openjdk.org/browse/JDK-8303942): FileMapInfo::write_bytes aborts on a short os::write

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13188/head:pull/13188` \
`$ git checkout pull/13188`

Update a local copy of the PR: \
`$ git checkout pull/13188` \
`$ git pull https://git.openjdk.org/jdk.git pull/13188/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13188`

View PR using the GUI difftool: \
`$ git pr show -t 13188`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13188.diff">https://git.openjdk.org/jdk/pull/13188.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13188#issuecomment-1492499162)

### CODE REVIEW

File: src/hotspot/share/cds/filemap.cpp [hunk 1]
```diff
@@ -1764,12 +1764,17 @@ size_t FileMapInfo::write_heap_regions(GrowableArray<MemRegion>* regions,
 
 void FileMapInfo::write_bytes(const void* buffer, size_t nbytes) {
   assert(_file_open, "must be");
-  ssize_t n = os::write(_fd, buffer, (unsigned int)nbytes);
-  if (n < 0 || (size_t)n != nbytes) {
-    // If the shared archive is corrupted, close it and remove it.
-    close();
-    remove(_full_path);
-    fail_stop("Unable to write to shared archive file.");
+  ssize_t size = (ssize_t)nbytes;
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/filemap.cpp:1767]: I'm very confused as to how the underlying OS `write` can take in an unsigned count of bytes to write, yet the actual count of what was written is signed and so cannot reflect the true count for large values! Can we at least assert here that size is > 0 ?

**Review by dholmes-ora — APPROVED** [review]
Seems correct - just one minor request below. Thanks.

**Comment by iklam** [comment]
We have other cases that assumes that `os::write` writes all the requested bytes:

https://github.com/openjdk/jdk/blob/aa762102e9328ca76663b56b3be6f6141b044744/src/hotspot/share/jfr/recorder/repository/jfrEmergencyDump.cpp#L379-L389

Writing a loop for each case will be messy and buggy. It's better to get rid of `os::write()` and replace it with `os::write_fully()`.

```
size_t bytes_written = os::write_fully(fd, buffer, bytes_requested);
if (bytes_written != bytes_requested) {
    // report error ....
}
```

and I can't understand why `os::write` takes an `unsigned int`!

https://github.com/openjdk/jdk/blob/aa762102e9328ca76663b56b3be6f6141b044744/src/hotspot/os/posix/os_posix.cpp#L774-L778

File: src/hotspot/share/cds/filemap.cpp (hunk 1 above)
Comment (afshin-zafari) [reply src/hotspot/share/cds/filemap.cpp:1767]: Can this be the rationale for it?
The return value can be `-1` so the type is `ssize_t`. And it is also large enough to hold the maximum bytes transfer of `write()`:
From the [write() man page](https://man7.org/linux/man-pages/man2/write.2.html#:~:text=of%20bytes%20written.-,On%20Linux%2C,-write()%20(and%20similar)) 
```
       On Linux, write() (and similar system calls) will transfer at
       most 0x7ffff000 (2,147,479,552) bytes, returning the number of
       bytes actually transferred.  (This is true on both 32-bit and
       64-bit systems.)

```

**Comment by afshin-zafari** [comment]
> We have other cases that assumes that `os::write` writes all the requested bytes:
> 
> https://github.com/openjdk/jdk/blob/aa762102e9328ca76663b56b3be6f6141b044744/src/hotspot/share/jfr/recorder/repository/jfrEmergencyDump.cpp#L379-L389
> 
> Writing a loop for each case will be messy and buggy. It's better to get rid of `os::write()` and replace it with `os::write_fully()`.
> 
> ```
> size_t bytes_written = os::write_fully(fd, buffer, bytes_requested);
> if (bytes_written != bytes_requested) {
>     // report error ....
> }
> ```
> 
> and I can't understand why `os::write` takes an `unsigned int`!
> 
> https://github.com/openjdk/jdk/blob/aa762102e9328ca76663b56b3be6f6141b044744/src/hotspot/os/posix/os_posix.cpp#L774-L778

For the `signed` and `unsigned` question, please refer to my reply to David's comment above. 

All cases that found are:

#### using `write()` in loop:
  src/hotspot/os/posix/perfMemory_posix.cpp
  src/hotspot/share/jfr/writers/jfrStreamWriterHost.inline.hpp
  src/hotspot/share/services/heapDumperCompression.cpp

#### buggy use of `write()`:
  src/hotspot/share/jfr/recorder/repository/jfrEmergencyDump.cpp
  src/hotspot/share/prims/jvm.cpp (Issue 8303943 is already created for it)
  src/hotspot/share/cds/filemap.cpp (The current issue is for it)

So, shall we create a new issue for the remaining buggy case and then create another issue for replacing all the loop cases with a new `write_fully()`? Or do all these in one new issue?

**Comment by dholmes-ora** [comment]
@iklam the JFR code you linked already handles the short write case. I only found two cases where it was not handled and bugs were filed for each - this is one of them. See JDK-8303937 for links.

**Comment by dholmes-ora** [comment]
> and I can't understand why os::write takes an unsigned int!

It is using the win32 signature. I'm going to assume at some point in time the Windows compiler was unhappy with casting size_t to unsigned int, so we made it take unsigned int.

File: src/hotspot/share/cds/filemap.cpp (hunk 1 above)
Comment (afshin-zafari) [reply src/hotspot/share/cds/filemap.cpp:1767]: `size` is checked to be `> 0` in loop's condition. Is assert for it still needed?

Comment (dholmes-ora) [reply src/hotspot/share/cds/filemap.cpp:1767]: With the new information an assert that nbytes <= SSIZE_MAX might still be appropriate. Let's see what others think.

**Comment by iklam** [comment]
> @iklam the JFR code you linked already handles the short write case. I only found two cases where it was not handled and bugs were filed for each - this is one of them. See JDK-8303937 for links.

Ths JFR code has multiple problems. 

```
// os.hpp
static ssize_t os::write(int fd, const void *buf, unsigned int nBytes);

// jfrEmergencyDump.cpp
int64_t bytes_read = 0;
int64_t bytes_written = 0;
...
bytes_written += (int64_t)os::write(emergency_fd, copy_block, bytes_read - bytes_written);
assert(bytes_read == bytes_written, "invariant");
```

(1) the casting of `int64_t` to `unsigned int nBytes` is dubious.
(2) it doesn't handle a `-1` result from `os::write`
(3) the `assert` is wrong.

I think this shows that it's very tricky to make a loop to write all the bytes requested. In particular, Most callers would have a `size_t` for the number of bytes to write. Asking everyone to consider between `ssize_t` vs `size_t` would be a maintenance nightmare:

```
ssize_t write(int fd, const void *buf, size_t count);

// The number of bytes written may be less than count if, for 
// example, there is insufficient space on the  underlying physical
// medium, or the RLIMIT_FSIZE resource limit is encountered 
// (see setrlimit(2)), or the call was interrupted by a signal handler
// after having written less than count bytes.  (See also pipe(7).)
```

Therefore, we should consolidate all the loops in a single place (preferably in platform-independent os.cpp and not once per platform).

```
// writes count bytes from the buffer starting at buf to the file referred to by the
// file descriptor fd.
// On  success,  0 is returned.  On error, -1 is returned, and errno is set to
// indicate the cause of the error.
int os::write_fully(int fd, const void *buf, size_t count);
```

For reference, see the Java `OutputStream.write()` call, which either writes all the bytes requested, or throws an IOException. 25+ years of Java shows that no one wants or needs to handle the partial writes:

- https://docs.oracle.com/javase/8/docs/api/java/io/OutputStream.html

**Comment by coleenp** [comment]
Until last week when I read these bugs, I didn't realize os::write() could not write the entire requested size.  I'd like os::write() to do that and if there's ever a case of needing a partial write, add a separately awkwardly named os::write_partial().  os::write should be the non-exceptional case so it's not used by mistake if there ever was the exceptional case needed.  Maybe use os::write_partial() for the JFR case and file a separate bug for that to use os::write properly.

**Comment by vnkozlov** [comment]
> Until last week when I read these bugs, I didn't realize `os::write()` could not write the entire requested size. I'd like os::write() to do that and if there's ever a case of needing a partial write, add a separately awkwardly named os::write_partial(). os::write should be the non-exceptional case so it's not used by mistake if there ever was the exceptional case needed. Maybe use os::write_partial() for the JFR case and file a separate bug for that to use os::write properly.

I fully support this suggestion. I am using `os::write()` in my current project and I don't want any surprises from it!

**Comment by dholmes-ora** [comment]
I'd just like to point out that `os::write` is just our wrapper around the platform `write` method and it is the underlying method that can fail to write the full amount. If we want some additional utility methods to `write_fully` then that is fine, but not sure they really belong in `os::` namespace, nor is it what `os::write` itself should provide.

**Comment by vnkozlov** [comment]
Our shared `os` class has many methods which are not wrappers. I think it is fine place for our `write_fully()` as @iklam suggested.

**Comment by coleenp** [comment]
I see so many bugs in the future like the ones we have now because os::write() is assumed to be write_fully, and it's not obvious that you need to call this function 99% of the time.  This way, people need more things to look for in code reviews.

**Comment by coleenp** [comment]
> 25+ years of Java shows that no one wants or needs to handle the partial writes:

Absolutely!

**Comment by afshin-zafari** [comment]
A new PR https://github.com/openjdk/jdk/pull/13750, is created for changing the `os::write` itself to include a loop and fix the calls and other wrong uses of it in the code.

**Pull request closed without merging**